File: implementations/swift/check.swift
import Foundation

@main
struct Check {
    static func main() {
        let fileManager = FileManager.default
        var mismatches: [(String, String)] = []
        var count = 0

        do {
            let entries = try fileManager.contentsOfDirectory(at: cidsDir, includingPropertiesForKeys: [.isDirectoryKey])
                .sorted { $0.lastPathComponent < $1.lastPathComponent }
            for entry in entries {
                let resourceValues = try entry.resourceValues(forKeys: [.isDirectoryKey])
                if resourceValues.isDirectory == true {
                    continue
                }
                count += 1
                let actual = entry.lastPathComponent
                let content = try Data(contentsOf: entry)
                let expected = try computeCID(content)
                if actual != expected {
                    mismatches.append((actual, expected))
                }
            }
        } catch {
            fputs("Error: \(error)\n", stderr)
            exit(1)
        }

        if mismatches.isEmpty {
            print("All \(count) CID files match their contents.")
            return
        }

        print("Found CID mismatches:")
        for (actual, expected) in mismatches {
            print("- \(actual) should be \(expected)")
        }
        exit(1)
    }
}

File: implementations/haskell/check.hs
module Main where

import CID (cidsDir, computeCID)
import Control.Monad (filterM, forM, unless)
import Data.List (sort)
import qualified Data.ByteString as BS
import System.Directory (doesDirectoryExist, doesFileExist, listDirectory)
import System.Exit (exitFailure)
import System.FilePath ((</>))

checkFile :: FilePath -> IO (Maybe String)
checkFile name = do
  let path = cidsDir </> name
  content <- BS.readFile path
  let expected = computeCID content
  if expected == name
    then pure Nothing
    else pure $ Just (name ++ " should be " ++ expected)

main :: IO ()
main = do
  exists <- doesDirectoryExist cidsDir
  unless exists $ do
    putStrLn "cids directory not found."
    exitFailure

  entries <- sort <$> listDirectory cidsDir
  files <- filterM (doesFileExist . (cidsDir </>)) entries
  results <- forM files checkFile
  let mismatches = [msg | Just msg <- results]
      count = length files

  if null mismatches
    then putStrLn $ "All " ++ show count ++ " CID files match their contents."
    else do
      putStrLn "Found CID mismatches:"
      mapM_ putStrLn $ map ("- " ++) mismatches
      exitFailure
